### PR TITLE
[SPARK-27462][SQL] Enhance insert into hive table that could choose some columns in target table flexibly.

### PR DESCRIPTION
**Pull request by beliefer**
## What changes were proposed in this pull request?

ANSI 2003 standard contains syntax about `insert into` as follows:
```
insert into gja_test_spark select * from gja_test;
insert into gja_test_spark select key, value, other from gja_test;
insert into gja_test_spark(key, value, other) select key, value, other from gja_test;
insert into gja_test_spark(key, value, other) select * from gja_test;
insert into gja_test_spark(key, value) select value, other from gja_test;
insert into gja_test_spark(key, other) select value, other from gja_test;
insert into gja_test_spark(value, other) select value, other from gja_test;
```
Hive 1.2.0+ support the syntax about `insert into` as follows:
```
insert into gja_test_spark select * from gja_test;
insert into gja_test_spark select key, value, other from gja_test;
insert into gja_test_spark(key, value, other) select key, value, other from gja_test;
insert into gja_test_spark(key, value, other) select * from gja_test;
insert into gja_test_spark(key, value) select value, other from gja_test;
insert into gja_test_spark(key, other) select value, other from gja_test;
insert into gja_test_spark(value, other) select value, other from gja_test;
insert into table gja_test_partition partition (col2 = 1) select * from gja_test;
insert into table gja_test_partition(key, value, other) partition (col2 = 1) select key, value, other from gja_test;
insert into table gja_test_partition(key, other) partition (col2 = 1) select key, other from gja_test;
```
This means that Hive can support flexible selection of some columns in the target table as the target columns for data insertion. See Description of [HIVE-9481](https://issues.apache.org/jira/browse/HIVE-9481) .
But Spark SQL only support 
```
insert into gja_test_spark select * from gja_test;
insert into gja_test_spark select key, value, other from gja_test;
insert into table gja_test_partition partition (col2 = 1) select * from gja_test;
insert into table gja_test_partition partition (col2 = 1) select key, value, other from gja_test;
```

This PR will enhance the feature. Implemented steps as follows:

**First**
I checked the implement of `HiveOutputWriter`, it writes data according to the index(position) of columns which belongs target table. So I think should use this feature.

**Second**
I add the new syntax and pass `insertedCols` as target columns for data insertion.

**Thrid**
In order to fill in the columns that are not selected, construct empty expressions and fill them in empty index(position) in `Project.projectList`.

**Notice**
This PR only enhance `insert into` for Hive.

## How was this patch tested?

Exists UT and new added UT.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala [hunk 1]
```diff
@@ -209,6 +209,31 @@ class PlanParserSuite extends AnalysisTest {
...
+    val plan = table("t").select('a2, 'b2)
+    def insert(
+        insertCols: Option[Seq[String]],
+        partition: Map[String, Option[String]],
+        overwrite: Boolean = false,
+        ifPartitionNotExists: Boolean = false): LogicalPlan =
+      InsertIntoTable(table("s"), insertCols, partition, plan, overwrite, ifPartitionNotExists)
+
+    // Single inserts
+    assertEqual(s"insert into s(a1, b1) $sql",
```
Comment (wangyum) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala]: Please add more tests. Example:
```sql
insert into s(b1) select b2 from t;
insert into s(b1, a1) select b2, a2 from t;
insert into s(A1, B1) select a2, b2 from t;
insert into s(B1) select a2, b2 from t;
...
```

Comment (beliefer) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala]: > Please add more tests. Example:
> 
> ```sql
> insert into s(b1) select b2 from t;
> insert into s(b1, a1) select b2, a2 from t;
> insert into s(A1, B1) select a2, b2 from t;
> insert into s(B1) select a2, b2 from t;
> ...
> ```

OK, I have supplement some UT.

**Comment by beliefer** [comment]
ok to test

**Comment by beliefer** [comment]
@cloud-fan @gatorsmile @srowen @dongjoon-hyun 
Could you review this PR?

**Comment by beliefer** [comment]
@srowen @cloud-fan 
Could you help me review this PR? Thanks.

**Comment by beliefer** [comment]
@srowen Could you make this PR to test? Thanks.

**Comment by beliefer** [comment]
cc @cloud-fan

**Comment by beliefer** [comment]
retest please

**Comment by beliefer** [comment]
@srowen Could you help me to review this PR? No one is concerned about this PR. So I have to ask for your help.

**Comment by srowen** [comment]
This one I really don't feel I can review reliably. I can retrigger tests though.

**Comment by beliefer** [comment]
> This one I really don't feel I can review reliably. I can retrigger tests though.

Thanks a lot. You are too modest. Maybe we nee more cautious.

**Comment by beliefer** [comment]
@cloud-fan @gatorsmile @vanzin @gengliangwang 
Could you help me to review this PR?

**Comment by beliefer** [comment]
@maropu Could you review this PR and make jenkins to test? Thanks a lot!

**Comment by beliefer** [comment]
@dongjoon-hyun Could you review this PR?Thanks a lot!

**Comment by beliefer** [comment]
@HyukjinKwon Could you review this PR?Thanks a lot!

**Comment by beliefer** [comment]
@cloud-fan Could you review this PR?Thanks a lot!

**Comment by beliefer** [comment]
@srowen Could you help me to retrigger tests?

**Comment by beliefer** [comment]
@srowen  I remember the Jenkins tests is an automatic process when a new commit come. Could you set this by some way? Thanks a lot!

**Comment by srowen** [comment]
Jenkins add to whitelist

**Comment by beliefer** [comment]
@HyukjinKwon Could you help me to review this PR? Thanks!

**Comment by beliefer** [comment]
@dongjoon-hyun  I need some help. Could you review this PR? Thanks.

**Comment by beliefer** [comment]
@cloud-fan Could you help me review this PR? If you are very busy, help introduce a committer to give me a review.

**Comment by beliefer** [comment]
cc @dongjoon-hyun @HyukjinKwon @cloud-fan

**Comment by beliefer** [comment]
cc @vanzin

**Comment by beliefer** [comment]
cc @HyukjinKwon

**Comment by beliefer** [comment]
cc @dongjoon-hyun

**Comment by beliefer** [comment]
cc @cloud-fan

**Comment by beliefer** [comment]
cc @maropu @zsxwing @gatorsmile

**Comment by beliefer** [comment]
cc @dongjoon-hyun

**Comment by beliefer** [comment]
cc @gatorsmile

**Comment by beliefer** [comment]
cc @cloud-fan @gatorsmile

**Comment by beliefer** [comment]
cc @rxin @vanzin @cloud-fan

**Comment by cloud-fan** [comment]
Can you put more details in the PR description? What's the behavior of the new syntax and how do you implement it?

**Comment by beliefer** [comment]
> Can you put more details in the PR description? What's the behavior of the new syntax and how do you implement it?

I have put some details in the PR description. If there still missing some information,I will continue to supplement.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 2]
```diff
@@ -332,24 +334,58 @@ case class PreprocessTableInsertion(conf: SQLConf) extends Rule[LogicalPlan] {
...
+      val cols = insertedCols.get
+      val project = insert.query.asInstanceOf[Project]
+      val filledProjectList = ArrayBuffer.empty[NamedExpression]
+      var i = 0
+      tableCols.foreach { tableCol =>
+        if (cols.contains(tableCol)) {
+          filledProjectList += project.projectList(i)
+          i += 1
+        } else {
+          filledProjectList += Alias(Literal(null, NullType), "NULL")()
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: what if the column is not nullable in the target table?

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: > what if the column is not nullable in the target table?

Hive may not allow column is not nullable.

**Comment by beliefer** [comment]
@cloud-fan wenchen, help me to review this PR again, thanks!

**Comment by beliefer** [comment]
cc @gatorsmile @dongjoon-hyun @zsxwing @vanzin

**Comment by srowen** [comment]
@beliefer this is way too many pings. Look at the history. I would infer that there isn't support for this change.

**Comment by beliefer** [comment]
> @beliefer this is way too many pings. Look at the history. I would infer that there isn't support for this change.

Yes,I made many pings. I will wait a committer to review this PR. If committers really don't want to support this syntax, I want know the reason overall consideration and planning.Thank you very much, srowen. You have more attention to normal contributors.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala (hunk 2 above)
Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: @cloud-fan wenchen, help me to review this PR again, thanks!

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: is there any other considerations?

**Comment by HyukjinKwon** [comment]
@beliefer can you actually check if this is ANSI standard, rather than comparing to Hive?

**Comment by beliefer** [comment]
> @beliefer can you actually check if this is ANSI standard, rather than comparing to Hive?

Yes, it is also a ANSI standard. [https://en.wikipedia.org/wiki/Insert_(SQL)](https://en.wikipedia.org/wiki/Insert_(SQL))

**Comment by HyukjinKwon** [comment]
Adding @mgaido91 and @maropu since it seems ANSI compliance issue.

**Comment by beliefer** [comment]
> Adding @mgaido91 and @maropu since it seems ANSI compliance issue.

Thansks for your focus and promote.

**Comment by mgaido91** [comment]
yes, SQL compliance would be great on this. But I am not sure it makes sense to implement it only for Hive as data source. It would be very confusing that Spark behaves differently according to which is the output source. Maybe it is worth to discuss this on the mailing list in order to get feedbacks, and if we decide to go on, I think we should do this for all the datasources.

**Comment by beliefer** [comment]
@mgaido91 Thanks for your review here.This PR only implement for Hive. Because I want we could focus on this point. If Spark allows implement for other data source, I will open other PR to implement.
AFAIK, some data source support the syntax, such as Hive, MySql and Oracle. But some data source not support the syntax, such as ElasticSearch, Redis and so on.
So I tend to implement for Hive, JDBC, text, parquet, orc. Give the rights of implement for other data source to a third party, we can throw an unsupported exception default.

**Comment by mgaido91** [comment]
@beliefer it sounds reasonable, but this is the kind of thing which breaks an assumption: you use Spark, you can switch the source and the output without breaking anything. So I think this feature (which in general I like, as it brings us closer to ANSI SQL compliance) doesn't come for free: that's the reason why I'd prefer if the adoption or not of this is discussed more broadly in the community. Once we decide whether it is more the pros or the cons of this, we can eventually go ahead IMHO.

**Comment by beliefer** [comment]
@mgaido91 I approve your opinion this feature not free and need more broad discussion. Spark sql already exists this issue,  some syntax (e.g create hive table, create data source table, insert overwrite and so on.) is the as the syntax this PR to implement. In a production environment, user switch source will increase learning and maintenance costs.

**Comment by gatorsmile** [comment]
This requires the concept of the DEFAULT value when we create a table. Thus, we need to first change our schema definition before improving INSERT.

**Review by gatorsmile — CHANGES_REQUESTED** [review]
This requires the concept of the DEFAULT value when we create a table. Thus, we need to first change our schema definition before improving INSERT.

**Comment by beliefer** [comment]
> This requires the concept of the DEFAULT value when we create a table. Thus, we need to first change our schema definition before improving INSERT.

Thanks for your review first of all. AFAIK, @wangyum will change schema definition with `DEFAULT` value and Hive3.0+ could support it.
I have two suggestions about the development order between `DEFAULT` and improving INSERT.
First, improving INSERT before support `DEFAULT` value which is considered after  @wangyum's work.
Second, waiting @wangyum to support Hive3.0+ and `DEFAULT` value, then we improving INSERT.
Maybe there exists other better suggestion.

**Comment by beliefer** [comment]
@gatorsmile Although before Hive 3.0 that do not support `DEFAULT` value, `NULL` value is the `DEFAULT` value in fact.

**Comment by gatorsmile** [comment]
We are trying our best to reduce coupling with Hive. Having a native support of Default in schema specification in Spark is what we need.

**Comment by beliefer** [comment]
> We are trying our best to reduce coupling with Hive. Having a native support of Default in schema specification in Spark is what we need.

AFAIK, `CatalogTable` depend on `Hive` metastore. What you mean is to say support `DEFAULT` value not depend on `Hive` metastore?

**Comment by wangyum** [comment]
@beliefer Hive metastore is just an implementation of `ExternalCatalog`. We support different versions of Hive metastore. It cannot be said that the default value is supported when using Hive 3.1 metastore, while the default value is not supported when using Hive 1.2 metastore. One possible implementation is similar to bucket, which is stored in table properties:
https://github.com/apache/spark/blob/f9776e389215255dc61efaa2eddd92a1fa754b48/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L444-L459
Please correct me if I'm wrong.

**Comment by beliefer** [comment]
Stay this.

**Pull request closed without merging**